pkg/controllers/machinesetsync/machineset_sync_controller.go

```diff
@@ -309,31 +339,63 @@ func (r *MachineSetSyncReconciler) reconcileMAPIMachineSetToCAPIMachineSet(ctx c
 	newCAPIMachineSet.OwnerReferences = []metav1.OwnerReference{clusterOwnerRefence}
 	// Set the paused annotation on the new CAPI MachineSet, as we want to create it paused.
 	annotations.AddAnnotations(newCAPIMachineSet, map[string]string{capiv1beta1.PausedAnnotation: ""})
+}
+
+// deleteOutdatedCAPIInfraMachineTemplates deletes infra machine templates that have MAPI machine label of the current MachineSet and don't have the current computed hash.
+func (r *MachineSetSyncReconciler) deleteOutdatedCAPIInfraMachineTemplates(ctx context.Context, mapiMachineSet *machinev1beta1.MachineSet, newCAPIInfraMachineTemplate client.Object) error {
+	logger := log.FromContext(ctx)
 
-	if result, err := r.ensureCAPIInfraMachineTemplate(ctx, mapiMachineSet, newCAPIMachineSet, newCAPIInfraMachineTemplate, clusterOwnerRefence); err != nil {
-		return result, fmt.Errorf("unable to ensure CAPI infra machine template: %w", err)
+	if mapiMachineSet == nil || len(mapiMachineSet.Name) == 0 {
+		return errNilMachineAPIMachineSet
 	}
 
-	if result, err := r.createOrUpdateCAPIMachineSet(ctx, mapiMachineSet, capiMachineSet, newCAPIMachineSet); err != nil {
-		return result, fmt.Errorf("unable to ensure CAPI machine set: %w", err)
+	machineSetMAPILabelSelector := labels.SelectorFromSet(map[string]string{machineSetLabelKey: mapiMachineSet.Name})
+
+	if newCAPIInfraMachineTemplate == nil || len(newCAPIInfraMachineTemplate.GetName()) == 0 {
+		return errNilInfraMachineTemplate
 	}
 
-	return ctrl.Result{}, r.applySynchronizedConditionWithPatch(ctx, mapiMachineSet, corev1.ConditionTrue,
-		consts.ReasonResourceSynchronized, messageSuccessfullySynchronizedMAPItoCAPI, &mapiMachineSet.Generation)
+	notNewCAPIInfraMachineTemplateNameFieldSelector := fields.OneTermNotEqualSelector("metadata.name", newCAPIInfraMachineTemplate.GetName())
+
+	listOptions := []client.DeleteAllOfOption{
+		client.InNamespace(r.CAPINamespace),
+		client.MatchingFieldsSelector{Selector: notNewCAPIInfraMachineTemplateNameFieldSelector},
+		client.MatchingLabelsSelector{Selector: machineSetMAPILabelSelector},
+	}
+
+	infraMachineTemplate, _, err := initInfraMachineTemplateAndInfraClusterFromProvider(r.Platform)
+	if err != nil {
+		return fmt.Errorf("failed to get infrastructure machine template from platform type: %w", err)
+	}
+
+	if err = r.DeleteAllOf(ctx, infraMachineTemplate, listOptions...); err != nil {
```

Comment:
This delete is called always. While this does not cause an issue, it feels wrong.
I think we might still want to list first to have feedback on what is to be deleted. The logger below also prints always, regardless if anything was deleted.

Reply:
I agree, can we list, check for items in the list, and then call the delete when there is a non-zero length list please? Otherwise we are making API calls regularly that do nothing, which is bad practice 

And we can also avoid the log spam too!